pyphare/pyphare/pharesee/hierarchy.py

```diff
@@ -0,0 +1,162 @@
+from ..core import box as boxm
+import numpy as np
+
+class PatchData:
+    """
+    base class for FieldData and ParticleData
+    this class just factors common geometrical properties
+    """
+    def __init__(self, layout, quantity):
+        """
+        :param layout: a GridLayout representing the domain on which the data
+             is defined
+        :param quantity: ['field', 'particle']
+        """
+        self.quantity = quantity
+        self.box     = layout.box
+        self.origin   = layout.origin
+        self.layout   = layout
+
+
+
+
+class FieldData(PatchData):
+    """
+    Concrete type of PatchData representing a physical quantity
+    defined on a grid.
+    """
+    def __init__(self, layout, field_name, data):
+        """
+        :param layout: A GridLayout representing the domain on which data is defined
+        :param field_name: the name of the field (e.g. "Bx")
+        :param data: the dataset from which data can be accessed
+        """
+        super().__init__(layout, 'field')
+
+        self.layout = layout
+        self.field_name = field_name
+        self.dx = layout.dl[0]
+
+        centering = layout.centering["X"][field_name]
+        self._ghosts_nbr = layout.nbrGhosts(layout.interp_order, centering)
+        self.ghost_box = boxm.grow(layout.box, self._ghosts_nbr)
+
+        if centering == "primal":
+            self.size = self.ghost_box.size() + 1
+            offset = 0
+        else:
+            self.size = self.ghost_box.size()
+            offset = 0.5*self.dx
+
+        self.x = self.origin[0] - self._ghosts_nbr * self.dx + np.arange(self.size) * self.dx + offset
+        self.dataset = data
+
+
+
+class ParticleData(PatchData):
+    """
+    Concrete type of PatchData representing particles in a region
+    """
+    def __init__(self, layout, data):
+        """
+        :param layout: A GridLayout object representing the domain in which particles are
+        :param data: dataset containing particles
+        """
+        super().__init__(layout, 'particles')
+        self.dataset = data
+        if layout.interp_order == 1:
+            self._ghosts_nbr = 1
+        elif layout.interp_order == 2 or layout.interp_order == 3:
+            self._ghosts_nbr = 2
+        else:
+            raise RuntimeError("invalid interpolation order {}".format(layout.interp_order))
+        self.ghost_box = boxm.grow(layout.box, self._ghosts_nbr)
+
+
+
+class Patch:
+    """
+    A patch represents a hyper-rectangular region of space
+    """
+
+    def __init__(self, patch_datas):
+        """
+        :param patch_datas: a list of PatchData objects
+        these are assumed to "belong" to the Patch so to
+        share the same origin, mesh size and box.
+        """
+        pdata0 = list(patch_datas.values())[0] #0 represents all others
```

Comment:
ok it just means that when reading the code, it could feel peculiar to see we're using the element 0 of a collection to get the box, origin, dx of the patch, why 0 ? why not 1 ? what's so special about the 0th element ?
The comment helps saying it's because 0 represents all others. They share the same value so taking 0 is ok.
The docstring also explains it by saying the Patch assumes all patch_datas given in arguments belong to the same level (so share the same dx, same origin and box)